src/symmetric/tweak_hash_tree.rs

```diff
@@ -10,44 +11,41 @@ pub struct HashTree<TH: TweakableHash> {
     layers: Vec<Vec<TH::Domain>>,
 }
 
-/// Function to compute a hash-tree given the leafs as input.
-/// The number of leafs must be a power of two.
+/// Function to compute a hash-tree given the leafs hashes as input.
+/// The number of leafs hashes must be a power of two.
 pub fn build_tree<TH: TweakableHash>(
     parameter: &TH::Parameter,
-    leafs: &[&[TH::Domain]],
+    leafs_hashes: Vec<TH::Domain>,
 ) -> HashTree<TH> {
     // check that number of leafs is a power of two
     assert!(
-        leafs.len().is_power_of_two(),
+        leafs_hashes.len().is_power_of_two(),
         "Hash-Tree build_tree: Number of leafs should be power of two"
     );
 
-    let mut layer_size = leafs.len();
-    let mut layers: Vec<Vec<TH::Domain>> = Vec::new();
+    let mut layer_size = leafs_hashes.len();
+    let mut layers: Vec<Vec<TH::Domain>> = Vec::with_capacity(layer_size.ilog2() as usize + 1);
 
     // the bottom layer contains the individual hashes of all leafs
-    layers.push(Vec::new());
-    for i in 0..layer_size {
-        let tweak = TH::tree_tweak(0, i as u32);
-        let hash = TH::apply(parameter, &tweak, leafs[i]);
-        layers[0].push(hash);
-    }
+    layers.push(leafs_hashes);
 
     // now, we build each layer by hashing pairs in the previous layer
     let mut level: u8 = 1;
     while layer_size >= 2 {
-        // start a new layer
-        layers.push(Vec::new());
         // this new layer will have half the size
         layer_size = layer_size / 2;
-        for i in 0..layer_size {
-            let left_idx = 2 * i;
-            let right_idx = 2 * i + 1;
-            let tweak = TH::tree_tweak(level, i as u32);
-            let children = &layers[(level - 1) as usize][left_idx..=right_idx];
-            let parent = TH::apply(parameter, &tweak, children);
-            layers[level as usize].push(parent);
-        }
+        layers.push(
```

Comment:
Just to make sure readers do not miss the parallelization here:

```suggestion
        // parallelize the two to one compressions
        layers.push(
```